<?xml version="1.0" encoding="UTF-8" ?>
<launch>
    <arg name="robot" default="X1"/>
    <arg name="robots" default="$(arg robot)"/>
    <arg name="robot_type" default="x1"/>
    <arg name="points" default="points_slow"/>
    <arg name="points_min_dist" default="0.125"/>

    <include file="$(find subt_virtual)/launch/helpers/for.launch">
        <arg name="items" value="$(arg robots)"/>
        <arg name="exclude" value="TEAMBASE"/>
        <arg name="param" value="name:naex_planner/robot_frames/{x},value:{x}/base_footprint"/>
    </include>

    <node name="naex_planner" pkg="naex" type="planner"
          respawn="true" respawn_delay="1.0" output="screen">

        <!--
             5 0.087
            10 0.175
            15 0.262
            20 0.349
            25 0.436
            30 0.524
            35 0.611
            40 0.698
            45 0.785
        -->
        <rosparam subst_value="true">
            position_name: x
            normal_name: normal_x
            map_frame: subt
            robot_frame: $(arg robot)/base_footprint
            max_cloud_age: 2.0
            input_range: 15.0
            max_pitch: 0.611
            max_roll: 0.524
            neighborhood_knn: 32
            neighborhood_radius: 0.6
            normal_radius: 0.5

            max_nn_height_diff: 0.15
            min_points_obstacle: 1
            max_ground_diff_std: 0.06
            max_ground_abs_diff_mean: 0.06
            edge_min_centroid_offset: 0.4
            min_dist_to_obstacle: 0.0

            viewpoints_update_freq: 1.
            min_vp_distance: 2.
            max_vp_distance: 5.
            self_factor: 0.5
            path_cost_pow: 0.75
            planning_freq: 0.5
            random_start: false

            min_num_empty: 4
            min_empty_ratio: 2.0
            max_occ_counter: 7
            min_empty_cos: 0.216

            points_min_dist: $(arg points_min_dist)
            filter_robots: true

            <!-- Have some clearance defaults. -->
            clearance_radius: 0.6
            clearance_low: 0.1
            clearance_high: 0.7

            num_input_clouds: 1
            input_queue_size: 15
        </rosparam>
        <!-- Customize things for particular robots. -->
        <rosparam if="$(eval robot_type == 'dtr')" subst_value="true">
            max_cloud_age: 5.0
            max_ground_diff_std: 0.1
            max_ground_abs_diff_mean: 0.1
            clearance_radius: 0.7
            clearance_low: 0.15
            clearance_high: 0.7
        </rosparam>
        <rosparam if="$(eval robot_type == 'marv')" subst_value="true">
            clearance_radius: 0.6
            clearance_low: 0.15
            clearance_high: 0.7
        </rosparam>
        <rosparam if="$(eval robot_type == 'tradr')" subst_value="true">
            clearance_radius: 0.5
            clearance_low: 0.2
            clearance_high: 0.6
            max_cloud_age: 5.0
        </rosparam>
        <rosparam if="$(eval 'x1' in robot_type)" subst_value="true">
            clearance_radius: 0.6
            clearance_low: 0.1
            clearance_high: 0.7
            num_input_clouds: 5
        </rosparam>
        <rosparam if="$(eval robot_type == 'x2')" subst_value="true">
            clearance_radius: 0.5
            clearance_low: 0.1
            clearance_high: 0.5
        </rosparam>

        <!-- Services -->
        <remap from="get_plan" to="get_plan"/>

        <!-- Inputs -->
        <remap from="input_map" to="~input_map"/>  <!-- Don't use map input. -->

        <remap from="input_cloud_0" to="$(arg points)"/>

        <remap if="$(eval robot_type == 'explorer_x1')"
               from="input_cloud_1" to="front_rgbd/points_slow"/>
        <remap if="$(eval robot_type == 'explorer_x1')"
               from="input_cloud_2" to="left_rgbd/points_slow"/>
        <remap if="$(eval robot_type == 'explorer_x1')"
               from="input_cloud_3" to="right_rgbd/points_slow"/>
        <remap if="$(eval robot_type == 'explorer_x1')"
               from="input_cloud_4" to="rear_rgbd/points_slow"/>

        <!-- Outputs -->
        <!-- Collected robot viewpoints -->
        <remap from="viewpoints" to="viewpoints"/>
        <remap from="other_viewpoints" to="other_viewpoints"/>

        <!--
            Global and local maps: don't subscribe global if not needed,
            use local map for visualization.
        -->
        <remap from="map" to="map"/>
        <remap from="local_map" to="local_map"/>
        <!-- Points to update (graph, features, labels) -->
        <remap from="dirty_map" to="dirty_map"/>
        <!-- Added / removed points -->
        <remap from="map_diff" to="map_diff"/>

        <!-- Planned path -->
        <remap from="path" to="path"/>
    </node>
</launch>
